### PR TITLE
Redshift: upload and copy in background, delete s3 temporary object

### PR DESCRIPTION
**Pull request by hito4t**

### CODE REVIEW

File: embulk-output-redshift/src/main/java/org/embulk/output/redshift/RedshiftCopyBatchInsert.java [hunk 1]
```diff
@@ -86,28 +93,37 @@ public void flush() throws IOException, SQLException
     {
         File file = closeCurrentFile();  // flush buffered data in writer
 
-        // TODO multi-threading
-        new UploadAndCopyTask(file, batchRows, s3KeyPrefix + UUID.randomUUID().toString()).call();
-        new DeleteFileFinalizer(file).close();
+        UploadAndCopyTask task = new UploadAndCopyTask(file, batchRows, s3KeyPrefix + UUID.randomUUID().toString());
+        executorService.submit(task);
```
Comment (frsyuki) [inline embulk-output-redshift/src/main/java/org/embulk/output/redshift/RedshiftCopyBatchInsert.java]: I think error handling is necessary to stop a bulk import job appropriately when upload or COPY failed.
`ExecutorService#submit()` returns `Future<Void>` and `Future#get()` throws ExecutionException if the callable throws an exception.

File: embulk-output-redshift/src/main/java/org/embulk/output/redshift/RedshiftCopyBatchInsert.java [hunk 2]
```diff
@@ -86,28 +93,37 @@ public void flush() throws IOException, SQLException
...
     }
 
     @Override
     public void finish() throws IOException, SQLException
     {
         super.finish();
+
+        executorService.shutdown();
+        try {
+            executorService.awaitTermination(Long.MAX_VALUE, TimeUnit.SECONDS);
```
Comment (frsyuki) [inline embulk-output-redshift/src/main/java/org/embulk/output/redshift/RedshiftCopyBatchInsert.java]: I think Long.MAX_VALUE is too long. How about using 1 hour?

File: embulk-output-redshift/src/main/java/org/embulk/output/redshift/RedshiftCopyBatchInsert.java [hunk 3]
```diff
@@ -86,28 +93,37 @@ public void flush() throws IOException, SQLException
...
+        }
+
         logger.info("Loaded {} files.", fileCount);
     }
 
     @Override
     public void close() throws IOException, SQLException
     {
+        if (!executorService.isShutdown()) {
+            executorService.shutdown();
```
Comment (frsyuki) [inline embulk-output-redshift/src/main/java/org/embulk/output/redshift/RedshiftCopyBatchInsert.java]: when `finish()` causes exception during `executorService.awaitTermination`, flow comes here. `ExecutorService#shutdown()` unlikely works. So, we should call `#shutdownNow()`.

Comment (frsyuki) [inline embulk-output-redshift/src/main/java/org/embulk/output/redshift/RedshiftCopyBatchInsert.java]: It is recommended by [JavaDoc](http://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ExecutorService.html) to wait for a while after `#shutdownNow()`

**Comment by hito4t** [comment]
Thank you for the review!
I've updated the source.

**Comment by hito4t** [comment]
I've modified to execute COPY in a different thread from UPLOAD.

**Comment by lance-edmodo** [comment]
We would like to get Redshift loads as fast and clean as possible. Is there some reason this pull request has been ignored?

**Comment by hito4t** [comment]
I'm sorry, I've been too busy.

I'll merge this PR after resolving conflict.

**Comment by hito4t** [comment]
Test result
Redshift: dc1.large
Input data: 10,000,000 records CSV file (2.2GB)

Client(embulk): c3.large (2 vCPU)
before: 10.2s
after: 7.6s

Client(embulk): c3.2xlarge (8 vCPU)
before: 5.8s
after: 2.8s

**Comment by lance-edmodo** [comment]
Cool! Thanks, I'll check it out soon.

**Pull request merged**